bundles/org.openhab.binding.hdpowerview/src/main/java/org/openhab/binding/hdpowerview/internal/handler/HDPowerViewShadeHandler.java

```diff
@@ -596,4 +611,100 @@ private void doRefreshShade(RefreshKind kind) {
+
+        // predicate to filter the vane channel
+        final Predicate<Channel> vaneChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_VANE
+                .equals(c.getChannelTypeUID());
+
+        // predicate to filter the secondary channel
+        final Predicate<Channel> secChannelFilter = c -> HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY
+                .equals(c.getChannelTypeUID());
+
+        // current and required state of the vane channel
+        final boolean vaneChannelExisting = channelStream.anyMatch(vaneChannelFilter);
+        final boolean vaneChannelRequired = capabilities.usesVaneChannel();
+
+        final boolean secChannelExisting = channelStream.anyMatch(secChannelFilter);
+        final boolean secChannelRequired = capabilities.usesSecondaryChannel();
+
+        if ((vaneChannelExisting == vaneChannelRequired) && (secChannelExisting == secChannelRequired)) {
+            // nothing to do
+            return true;
+        }
+
+        // make a mutable copy of the original immutable channel list
+        channels = new ArrayList<>(channels);
+
+        boolean error = false;
+        if (vaneChannelRequired) {
+            // build the vane channel
+            Channel vaneChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(), HDPowerViewBindingConstants.CHANNEL_SHADE_VANE))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_VANE).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-vane.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-vane.label")).build();
+
+            // add the vane channel
+            error |= !channels.add(vaneChannel);
+        } else {
+            // remove the vane channel
+            error |= !channels.removeIf(vaneChannelFilter);
+        }
+
+        if (secChannelRequired) {
+            // build the secondary channel
+            Channel secChannel = ChannelBuilder
+                    .create(new ChannelUID(thing.getUID(),
+                            HDPowerViewBindingConstants.CHANNEL_SHADE_SECONDARY_POSITION))
+                    .withType(HDPowerViewBindingConstants.CHANNEL_TYPE_SECONDARY).withKind(ChannelKind.STATE)
+                    .withAcceptedItemType(CoreItemFactory.DIMMER)
+                    .withDescription(translater.getText("channel-type.hdpowerview.shade-secondary.description"))
+                    .withLabel(translater.getText("channel-type.hdpowerview.shade-secondary.label")).build();
+
+            // add the secondary channel
+            error |= !channels.add(secChannel);
+        } else {
+            // remove the secondary channel
+            error |= !channels.removeIf(secChannelFilter);
+        }
+
+        if (error) {
+            // something went wrong
+            return false;
```

Comment:
What will happen after this? The actual cause is interesting, because if this can actually happen, we could potentially get stuck in a loop trying again and again to construct these channels each time the shade is updated? I would assume it would fail consistently unless we have a concurrency issue.

Reply:
I think a) it will never happen, b) if it did happen then it would do so consistently, but c) we would not be be stuck in a loop, since updateCapabilities() will have only called us one time...   ...